### DISCUSSION TITLE
Extending primitives with custom validation methods (that are chainable & under the same export)

### DISCUSSION THREAD

**Discussion by DominikSerafin** [colinhacks/zod#516] [Q&A]
I'm creating a new discussion for this as I want to avoid spamming the previous discussion and also because I think it will be more discoverable/searchable for others that may have this issue as well. 

<details><summary>Original messages from <a href="https://github.com/colinhacks/zod/discussions/366">#366</a></summary>
<div>

>DominikSerafin:
>
> @colinhacks I've read multiple issues, PRs and discussions here in this repo, and somehow I still can't come up with a solution to simply create a new custom validation method for primitives that is chainable (in Zod v3).
> 
> What I'm basically trying to achieve is to be able to write something like this:
> 
> ```ts
> import * as z from './zod-extended';
> 
> const schema = z.object({
>   text: z.string().min(5).nonewlines().max(200),
> });
> const parsed = schema.parse({text: 'example'});
> ```
> 
> This example has `nonewlines()` method chained between other string methods (which should throw an error if string has any newline characters).
> 
> This is just an example, though, as I'd want to find a way to add a custom validation methods (with custom error code & message) for some of the other primitive types as well (string, number, bigint, date).
> 
> The re-export pattern doesn't let me chain my custom method. And I've tried the subclass approach, but I wasn't able to figure out how to do that properly (I've looked into `ZodString` implementation, like you've suggested).
> 
> I would really appreciate your help.

>colinhacks:
>
> Here's a simple example of implementing a subclass of ZodString. As long as all your methods delegate to `.refine` and `.transform`, this is fairly straightforward.
> 
> ```ts
> class CustomString extends z.ZodString {
>   toNumber() {
>     return this.transform((val) => val.length);
>   }
> 
>   static create = (): CustomString => {
>     return new CustomString({
>       checks: [],
>     });
>   };
> }
> 
> const mySchema = CustomString.create().toNumber();
> ```
> 
> If you want to modify the parsing logic, add additional state/properties, etc, things get more complicated. You'll just have to bite the bullet and learn how the existing schema types work.

</div>
</details>

---

@colinhacks, thank you for showing me the proper way to subclass `ZodString`. Unfortunately, I still can't make this work. 

This is my Zod abstraction `zod-extended.ts`:

```typescript
import {ZodString} from 'zod';

export * from 'zod';

class ZodStringExtended extends ZodString {
  nonewlines() {
    return this.refine((val) => !/\r?\n+/gi.test(val), {
      message: `String can't have newline characters`,
    });
  }

  static create = (): ZodStringExtended => {
    return new ZodStringExtended({
      checks: [],
    });
  };
}

const stringType = ZodStringExtended.create;
export {stringType as string};
```

This works when I'm doing something like this:

```typescript
import * as z from './zod-extended';
const schema = z.object({
  text: z.string().nonewlines(),
 );
const parsed = schema.parse({text: 'example'});
```

However, it doesn't let me chain my custom method with other `ZodString` methods.

```typescript
...
  text: z.string().nonewlines().url(), // Property 'url' does not exist on type 'ZodEffects<ZodStringExtended, string>'.
...
```

```typescript
...
  text: z.string().url().nonewlines(), // Property 'nonewlines' does not exist on type 'ZodString'.
...
```

---

BTW in [yup](https://github.com/jquense/yup) this is pretty straightforward using the [addMethod](https://github.com/jquense/yup#yupaddmethodschematype-schema-name-string-method--schema-void) helper. I wish Zod had something like this.

**Comment by VinSpee** [comment]
Did anyone ever come up with an answer to this? I'm struggling with it currently

**Reply by toxsick** [reply]
Me too, I would love to use chaining for my own custom validations, because otherwise my schema definitions do not look zod-like anymore. Any leads on how to do this would be much  appreciated.

BTW, thanks @DominikSerafin for the provided example and the writeup!

**Comment by ChromeGG** [comment]
I solved this issue in this way
```ts
// validation.ts
import { z as zod } from 'zod'

export const z = {
  ...zod,
  id: () => zod.number().int().min(0),
}
```

simple and readable, but I have one issue with `z.infer`: `Cannot find namespace 'z'.ts(2503)`. Any ideas on how to solve that?

**Reply by benknab** [reply]
I think you need to reexport the types also, you're only reexporting `z` the library, but not `z` the namespace

**Reply by jimmyn** [reply]
Anyone found the solution?

**Comment by mamlzy** [comment]
any updates?

**Reply by vladimir-gal** [reply]
```ts
import { z, ZodTypeAny } from 'zod'

declare module 'zod' {
  interface ZodString {
    noLeadingNumber(): ZodEffects<this, string, string>
    noTrailingNumber(): ZodEffects<this, string, string>
  }

  interface ZodEffects<
    T extends ZodTypeAny,
    Output = T['_output'],
    Input = T['_input']
  > {
    noLeadingNumber(): ZodEffects<this, Output, Input>
    noTrailingNumber(): ZodEffects<this, Output, Input>
  }
}

z.ZodString.prototype.noLeadingNumber = function () {
  return this.refine((val) => (/^\d/.test(val) ? false : true), {
    message: 'No leading number',
  })
}

z.ZodString.prototype.noTrailingNumber = function () {
  return this.refine((val) => (/\d$/.test(val) ? false : true), {
    message: 'No trailing number',
  })
}

z.ZodEffects.prototype.noLeadingNumber = function () {
  return this.refine((val) => !/^\d/.test(val), {
    message: 'No leading number',
  })
}

z.ZodEffects.prototype.noTrailingNumber = function () {
  return this.refine((val) => !/\d$/.test(val), {
    message: 'No trailing number',
  })
}
```

This just worked for me, you need to ensure you are extending both `ZodString` and `ZodEffect` so you can chain the calls.

Now something like this will work:
```ts
const schema = z.object({
  lastName: z.string().min(2).noLeadingNumber().noTrailingNumber(),
})
```

**Reply by MaLiN2223** [reply]
Annoyingly it seems that there's no way to return ZodString from any of those, so you end up being unable to chain ZodString methods after any of the ones you listed. 
e.g.

```
const schema = z.object({
  lastName: z.string().noLeadingNumber().noTrailingNumber().min(2),
})
```

won't work.

**Reply by RobinTail** [reply]
@MaLiN2223 , this is because `.refine()` (used by `.noTrailingNumber()`) returns `ZodEffects`.
In order to make it work, you'd either:
- move `.min(2)` next to `z.string()`;
- or make another `.refine((str) => str.length >= 2)` if you need this check in the end of the sequence

**Reply by MaLiN2223** [reply]
Yes, indeed.
I'm using a codegen which puts all functions in a specific order so it wasn't feasible for me to reorder the calls.
I ended up extending ZodEffect with `min` (which is essentially the refine you mentioned)

**Comment by RobinTail** [comment]
Here is the approach for extending Zod [recommended by Colin](https://github.com/colinhacks/zod/pull/3445#issuecomment-2091463120):

https://github.com/colinhacks/zod/blob/90efe7fa6135119224412c7081bd12ef0bccef26/plugin/effect/src/index.ts#L21-L31
